File: test-suite/csharp/security/ArchiveExtractionClean.cs
using System;
using System.IO;
using System.IO.Compression;
using System.Formats.Tar;
using SharpCompress.Archives;

public static class ArchiveExtractionClean
{
    private static string GetSafeExtractionPath(string destination, string entryName)
    {
        var destinationRoot = Path.GetFullPath(destination);
        var target = Path.GetFullPath(Path.Combine(destinationRoot, entryName));
        if (!target.StartsWith(destinationRoot + Path.DirectorySeparatorChar, StringComparison.Ordinal) &&
            !string.Equals(target, destinationRoot, StringComparison.Ordinal))
        {
            throw new InvalidOperationException("Archive entry escapes destination");
        }

        return target;
    }

    public static void UnzipSafely(ZipArchive archive, string destination)
    {
        foreach (var entry in archive.Entries)
        {
            var target = GetSafeExtractionPath(destination, entry.FullName);
            Directory.CreateDirectory(Path.GetDirectoryName(target)!);
            entry.ExtractToFile(target, overwrite: true);
        }
    }

    public static void UntarSafely(TarReader reader, string destination)
    {
        while (reader.GetNextEntry() is { } entry)
        {
            var target = Path.GetFullPath(Path.Combine(destination, entry.Name));
            var root = Path.GetFullPath(destination);
            if (!target.StartsWith(root + Path.DirectorySeparatorChar, StringComparison.Ordinal))
            {
                throw new InvalidOperationException("Archive entry escapes destination");
            }

            using var output = File.OpenWrite(target);
            entry.DataStream?.CopyTo(output);
        }
    }

    public static void ExtractSharpCompressEntry(IArchiveEntry entry, string destination)
    {
        var target = GetSafeExtractionPath(destination, entry.Key);
        entry.WriteToFile(target);
    }
}

File: test-suite/java/security/ArchiveExtractionClean.java
package security;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

final class ArchiveExtractionClean {
    void unzipSafely(Path destination, ZipInputStream archive) throws IOException {
        Path base = destination.toRealPath().normalize();
        ZipEntry entry;
        while ((entry = archive.getNextEntry()) != null) {
            Path target = safeDestination(base, entry.getName());
            if (entry.isDirectory()) {
                Files.createDirectories(target);
                continue;
            }
            Files.createDirectories(target.getParent());
            try (OutputStream output = Files.newOutputStream(target)) {
                archive.transferTo(output);
            }
        }
    }

    void unzipWithInlineCheck(Path destination, ZipInputStream archive) throws IOException {
        Path base = destination.toRealPath().normalize();
        ZipEntry entry;
        while ((entry = archive.getNextEntry()) != null) {
            Path target = base.resolve(entry.getName()).normalize();
            if (!target.startsWith(base)) {
                throw new IOException("archive entry escapes destination");
            }
            if (entry.isDirectory()) {
                Files.createDirectories(target);
                continue;
            }
            Files.copy(archive, target);
        }
    }

    private Path safeDestination(Path base, String entryName) throws IOException {
        Path target = base.resolve(entryName).normalize();
        if (!target.startsWith(base)) {
            throw new IOException("archive entry escapes destination");
        }
        return target;
    }
}
